DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/board_test/src/usm_speed.hpp

```diff
@@ -0,0 +1,187 @@
+                        size_t num_items) {
+  return q.single_task<USMMemRead>([=]() {
+    sycl::host_ptr<sycl::ulong8> in_h(in);
+    sycl::host_ptr<sycl::ulong8> out_h(out);
+    sycl::ulong8 sum{0};
+    for (size_t i = 0; i < num_items; i++) {
+      sum += in_h[i];
+    }
+    // This prevents the reads from being optimized away
+    out_h[0] = sum;
+  });
+}
+
+// WRITE test: launches a kernel to write data to a USM pointer.
+sycl::event write_kernel(sycl::queue &q, sycl::ulong8 *in, sycl::ulong8 *out,
+                         size_t num_items) {
+  return q.single_task<USMMemWrite>([=]() {
+    sycl::host_ptr<sycl::ulong8> out_h(out);
+    sycl::ulong8 answer{TEST_VAL};
+    for (size_t i = 0; i < num_items; i++) {
+      out_h[i] = answer;
+    }
+  });
+}
+
+// Function to check output against expected answer.
+bool verify(sycl::ulong8 *actual, sycl::ulong8 *expected, size_t num_items) {
+  // Verify all values are equal
+  for (int i = 0; i < num_items; i++) {
+    for (int j = 0; j < 8; j++) {
+      if (actual[i][j] != expected[i][j]) {
+        std::cerr << "ERROR: Values do not match, "
+                  << "out[" << i << "][" << j << "] = " << actual[i][j]
+                  << "; expected " << expected[i][j] << std::endl;
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+// Parameterized function to perform one of the tests defined above. Allocates
+// and initializes host USM, runs the test, then verifies the answer. Then
+// re-runs the test several times to measure bandwidth.
+int run_test(sycl::queue &q, USMTest test) {
+
+  size_t iterations = 1;
+  size_t num_bytes = 1024 * 1024 * 1024;
+  size_t num_items = num_bytes / sizeof(sycl::ulong8);
+  size_t num_bytes_GB = num_bytes / kGB;
+
+  std::cout << "Iterations: " << iterations << std::endl;
+  std::cout << "Data size: " << num_bytes / kMB << " MB" << std::endl;
+  std::cout << "Data type size: " << sizeof(sycl::ulong8) << " bytes"
+            << std::endl;
+
+  // USM host allocation
+  sycl::ulong8 *in = sycl::malloc_host<sycl::ulong8>(num_items, q);
+  sycl::ulong8 *out = sycl::malloc_host<sycl::ulong8>(num_items, q);
+  if (in == nullptr || out == nullptr) {
+    std::cerr << "Error: Out of memory, can't allocate " << num_bytes
+              << " bytes" << std::endl;
+    return 1;
+  }
+
+  // Initialize the input with random values and output with zero.
+  for (int i = 0; i < num_items; i++) {
+    in[i] = rand();
+    out[i] = 0;
+  }
+
+  std::function<sycl::event(sycl::queue &, sycl::ulong8 *, sycl::ulong8 *,
+                            size_t)>
+      kernel;
+  sycl::ulong8 *expected = new sycl::ulong8[num_items];
+
+  // Test selection: set the test function that will be run, and fill the
+  // "expected" array with the expected answer.
+  switch (test) {
+  case MEMCOPY: {
+    std::cout << "Case: Full Duplex" << std::endl;
+    kernel = memcopy_kernel;
+    // Full duplex transfers twice the amount of data
+    num_bytes_GB *= 2;
```

Comment:
I don't think you should be doubling the num_bytes_GB. e.g. 100Mbps full-duplex means 100 Mbps each way, not 200Mbps.

the same amount of data is transferred; just that it is transferred in, and then out again.

you will need to update this in the readme too.